README.md
# GraphQL Koa Middleware

[![License][license_badge]][license] [![Pipelines][pipelines_badge]][pipelines] [![NPM][npm_badge]][npm] [![semantic-release][semantic_release_badge]][semantic_release]

This is a fork from [koa-graphql] and [express-graphql]. Since koa-graphql does not seems to update its package.json to allow `"graphql": "^15.0.0"`, I forked it.

## Getting Started

```
npm i koa-graphql-typescript
```

It is a drop-in replacement for koa-graphql except no graphiql included.

```typescript
import Koa from "koa";
import Router from "koa-router";
import { graphqlHTTP } from "koa-graphql-typescript";

const app = new Koa();
const router = new Router();

router.post("/graphql", graphqlHTTP({ schema: MyGraphQLSchema }));

app.use(router.routes()).use(router.allowedMethods());
```

## Differences

There are some differences between koa-graphql-typescript and koa-graphql.

### No Longer Depends On express-graphql

koa-graphql is somewhat between an adapter for express-graphql and being its own library. If it is adapter and it converts Koa request into Express request, it can get the latest updates from express-graphql.

However, koa-graphql has its own dependencies and logic on top of express-graphql. It will be better off on its own.

### No Longer Include graphiql

There are many Graphql clients available, including server-side and standalone clients. There is not need to include a client that it may not be used and its dependencies.

If you want to provide a server-side client on the same endpoint, you can provide it via `router.get` because graphql endpoint only uses POST.

### Rewritten In TypeScript Instead Of Flow

I am more familiar with TypeScript. Also, the popularity of TypeScript is far superior than Flow. But this should not matter if you just use the library.

[license]: ./LICENSE
[license_badge]: https://img.shields.io/badge/license-Apache--2.0-green.svg
[pipelines]: https://github.com/joshuaavalon/koa-graphql-typescript/actions/workflows/main.yml
[pipelines_badge]: https://github.com/joshuaavalon/koa-graphql-typescript/actions/workflows/main.yml/badge.svg
[npm]: https://www.npmjs.com/package/koa-graphql-typescript
[npm_badge]: https://img.shields.io/npm/v/koa-graphql-typescript/latest.svg
[semantic_release]: https://github.com/semantic-release/semantic-release
[semantic_release_badge]: https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg
[koa-graphql]: https://github.com/graphql-community/koa-graphql
[express-graphql]: https://github.com/graphql/express-graphql


CHANGELOG.md
## [1.0.4](https://github.com/joshuaavalon/koa-graphql-typescript/compare/v1.0.3...v1.0.4) (2021-03-20)


### Bug Fixes

* undefined empty body ([45e3a5e](https://github.com/joshuaavalon/koa-graphql-typescript/commit/45e3a5e0309d4fa48e18a2be6f79409637772f3c))

## [1.0.3](https://github.com/joshuaavalon/koa-graphql-typescript/compare/v1.0.2...v1.0.3) (2021-03-20)


### Bug Fixes

* change to CommonJS ([d3e4bb5](https://github.com/joshuaavalon/koa-graphql-typescript/commit/d3e4bb5b597b470e7035d9d0141cebc22d15e7b5))

## [1.0.2](https://github.com/joshuaavalon/koa-graphql-typescript/compare/v1.0.1...v1.0.2) (2021-03-20)


### Bug Fixes

* package.json files ([59332c5](https://github.com/joshuaavalon/koa-graphql-typescript/commit/59332c5e829f7fdc7fe548cff11185b1fd960e9c))

## [1.0.1](https://github.com/joshuaavalon/koa-graphql-typescript/compare/v1.0.0...v1.0.1) (2021-03-20)


### Bug Fixes

* add missing build ([d7ec741](https://github.com/joshuaavalon/koa-graphql-typescript/commit/d7ec7415b86bace028132eac83697842972e7ed2))

# 1.0.0 (2021-03-20)


### Features

* initial commit ([2e112c2](https://github.com/joshuaavalon/koa-graphql-typescript/commit/2e112c22b4d5ea1bd126f7a995e3ecfe202307a8))


src/index.ts
import { formatError } from "graphql";
import httpError from "http-errors";

import { getGraphQLParams } from "./graphql-param";
import { Options, parseOptions } from "./option";
import { parseQuery } from "./parse-query";

import type { Middleware } from "koa";
import type { ExecutionResult } from "graphql";

export const graphqlHTTP = (options: Options): Middleware => {
  if (!options) {
    throw new Error("GraphQL middleware requires options.");
  }
  return async (ctx, next) => {
    const { request, response, req, body } = ctx;

    // GraphQL HTTP only supports GET and POST methods.
    if (request.method !== "GET" && request.method !== "POST") {
      response.set("Allow", "GET, POST");
      throw httpError(405, "GraphQL only supports GET and POST requests.");
    }

    const data = await parseOptions(options, ctx);
    const {
      schema,
      formatError: formatErrorFn,
      extensions: extensionsFn,
      pretty
    } = data;

    // Assert that schema is required.
    if (!schema) {
      throw new Error("GraphQL middleware options must contain a schema.");
    }

    let result: ExecutionResult;
    try {
      // Parse the Request to get GraphQL request parameters.
      const params = await getGraphQLParams(req, body);
      const { variables, operationName } = params;

      const query = await parseQuery(data, params, ctx);
      result = query.result;

      // Collect and apply any metadata extensions if a function was provided.
      // http://facebook.github.io/graphql/#sec-Response-Format
      if (result && extensionsFn) {
        const extensions = extensionsFn({
          document: query.document,
          variables,
          operationName,
          result,
          context: ctx
        });
        if (extensions && typeof extensions === "object") {
          result.extensions = extensions;
        }
      }
    } catch (error) {
      // If an error was caught, report the httpError status, or 500.
      response.status = error.status || 500;
      result = { errors: [error] };
    }

    // If no data was included in the result, that indicates a runtime query
    // error, indicate as such with a generic status code.
    // Note: Information about the error itself will still be contained in
    // the resulting JSON payload.
    // http://facebook.github.io/graphql/#sec-Data
    if (response.status === 200 && result && !result.data) {
      response.status = 500;
    }
    // Format any encountered errors.
    if (result && result.errors) {
      result.errors = result.errors.map(err =>
        formatErrorFn ? formatErrorFn(err, ctx) : formatError(err)
      );
    }

    // Otherwise, present JSON directly.
    const payload = pretty ? JSON.stringify(result, null, 2) : result;
    response.type = "application/json";
    response.body = payload;
    await next();
  };
};


src/option.ts
import type { Context, Request, Response } from "koa";
import type {
  ASTVisitor,
  GraphQLError,
  GraphQLFieldResolver,
  GraphQLSchema,
  ValidationContext
} from "graphql";

import type { MayBePromise, RequestInfo } from "./type";

export type Options =
  | ((request: Request, response: Response, context: Context) => OptionsResult)
  | OptionsResult;
export type OptionsResult = OptionsData | Promise<OptionsData>;
export type OptionsData = {
  /**
   * A GraphQL schema from graphql-js.
   */
  schema: GraphQLSchema;

  /**
   * An object to pass as the rootValue to the graphql() function.
   */
  rootValue?: any;

  /**
   * A boolean to configure whether the output should be pretty-printed.
   */
  pretty?: boolean;

  /**
   * An optional function which will be used to format any errors produced by
   * fulfilling a GraphQL operation. If no function is provided, GraphQL's
   * default spec-compliant `formatError` function will be used.
   */
  formatError?: (error: GraphQLError, context?: Context) => any;

  /**
   * An optional array of validation rules that will be applied on the document
   * in additional to those defined by the GraphQL spec.
   */
  validationRules?: Array<(ctx: ValidationContext) => ASTVisitor>;

  /**
   * An optional function for adding additional metadata to the GraphQL response
   * as a key-value object. The result will be added to "extensions" field in
   * the resulting JSON. This is often a useful place to add development time
   * info such as the runtime of a query or the amount of resources consumed.
   *
   * Information about the request is provided to be used.
   *
   * This function may be async.
   */
  extensions?: (info: RequestInfo) => MayBePromise<{ [key: string]: any }>;

  /**
   * A resolver function to use when one is not provided by the schema.
   * If not provided, the default field resolver is used (which looks for a
   * value or method on the source value with the field's name).
   */
  fieldResolver?: GraphQLFieldResolver<any, any>;
};

export const parseOptions = async (
  opts: Options,
  ctx: Context
): Promise<OptionsData> => {
  const { request, response } = ctx;
  // Resolve the Options to get OptionsData.
  const optionsData = await (typeof opts === "function"
    ? opts(request, response, ctx)
    : opts);

  // Assert that optionsData is in fact an Object.
  if (!optionsData || typeof optionsData !== "object") {
    throw new Error(
      "GraphQL middleware option function must return an options object " +
        "or a promise which will be resolved to an options object."
    );
  }
  return optionsData;
};


src/parseBody.ts
import type { IncomingMessage } from "http";
import type { Gunzip, Inflate } from "zlib";
import zlib from "zlib";
import querystring from "querystring";
import { URLSearchParams } from "url";
import getBody from "raw-body";
import httpError from "http-errors";
import contentType from "content-type";

import type { ParsedMediaType } from "content-type";

export interface GraphQLParams {
  query: string | null;
  variables: { readonly [name: string]: unknown } | null;
  operationName: string | null;
  raw: boolean;
}

export type Request = IncomingMessage & { body?: unknown };

// Return a decompressed stream, given an encoding.
function decompressed(
  req: Request,
  encoding: string
): Request | Inflate | Gunzip {
  switch (encoding) {
    case "identity":
      return req;
    case "deflate":
      return req.pipe(zlib.createInflate());
    case "gzip":
      return req.pipe(zlib.createGunzip());
  }
  throw httpError(415, `Unsupported content-encoding "${encoding}".`);
}

/**
 * RegExp to match an Object-opening brace "{" as the first non-space
 * in a string. Allowed whitespace is defined in RFC 7159:
 *
 *     ' '   Space
 *     '\t'  Horizontal tab
 *     '\n'  Line feed or New line
 *     '\r'  Carriage return
 */
const jsonObjRegex = /^[ \t\n\r]*\{/u;

// Read and parse a request body.
async function readBody(
  req: Request,
  typeInfo: ParsedMediaType
): Promise<string> {
  const charset = typeInfo.parameters.charset?.toLowerCase() ?? "utf-8";

  // Assert charset encoding per JSON RFC 7159 sec 8.1
  if (!charset.startsWith("utf-")) {
    throw httpError(415, `Unsupported charset "${charset.toUpperCase()}".`);
  }

  // Get content-encoding (e.g. gzip)
  const contentEncoding = req.headers["content-encoding"];
  const encoding =
    typeof contentEncoding === "string"
      ? contentEncoding.toLowerCase()
      : "identity";
  const length = encoding === "identity" ? req.headers["content-length"] : null;
  const limit = 100 * 1024; // 100kb
  const stream = decompressed(req, encoding);

  // Read body from stream.
  try {
    return await getBody(stream, { encoding: charset, length, limit });
  } catch (rawError: unknown) {
    const error = httpError(
      400,
      /* istanbul ignore next: Thrown by underlying library. */
      rawError instanceof Error ? rawError : String(rawError)
    );

    error.message =
      error.type === "encoding.unsupported"
        ? `Unsupported charset "${charset.toUpperCase()}".`
        : `Invalid body: ${error.message}.`;
    throw error;
  }
}

/**
 * Provided a "Request" provided by express or connect (typically a node style
 * HTTPClientRequest), Promise the body data contained.
 */
export async function parseBody(
  req: Request
): Promise<{ [param: string]: unknown }> {
  const { body } = req;

  // If express has already parsed a body as a keyed object, use it.
  if (typeof body === "object" && !(body instanceof Buffer)) {
    return body as { [param: string]: unknown };
  }

  // Skip requests without content types.
  if (req.headers["content-type"] === undefined) {
    return {};
  }

  const typeInfo = contentType.parse(req);

  // If express has already parsed a body as a string, and the content-type
  // was application/graphql, parse the string body.
  if (typeof body === "string" && typeInfo.type === "application/graphql") {
    return { query: body };
  }

  // Already parsed body we didn't recognise? Parse nothing.
  if (body !== null) {
    return {};
  }

  const rawBody = await readBody(req, typeInfo);
  // Use the correct body parser based on Content-Type header.
  switch (typeInfo.type) {
    case "application/graphql":
      return { query: rawBody };
    case "application/json":
      if (jsonObjRegex.test(rawBody)) {
        try {
          return JSON.parse(rawBody);
        } catch {
          // Do nothing
        }
      }
      throw httpError(400, "POST body sent invalid JSON.");
    case "application/x-www-form-urlencoded":
      return querystring.parse(rawBody);
  }

  // If no Content-Type header matches, parse nothing.
  return {};
}

export async function getGraphQLParams(
  request: Request
): Promise<GraphQLParams> {
  const urlData = new URLSearchParams(request.url?.split("?")[1]);
  const bodyData = await parseBody(request);

  // GraphQL Query string.
  let query = urlData.get("query") ?? (bodyData.query as string | null);
  if (typeof query !== "string") {
    query = null;
  }

  // Parse the variables if needed.
  let variables = (urlData.get("variables") ?? bodyData.variables) as {
    readonly [name: string]: unknown;
  } | null;
  if (typeof variables === "string") {
    try {
      variables = JSON.parse(variables);
    } catch {
      throw httpError(400, "Variables are invalid JSON.");
    }
  } else if (typeof variables !== "object") {
    variables = null;
  }

  // Name of GraphQL operation to execute.
  let operationName =
    urlData.get("operationName") ?? (bodyData.operationName as string | null);
  if (typeof operationName !== "string") {
    operationName = null;
  }

  const raw = urlData.get("raw") !== null || bodyData.raw !== undefined;

  return { query, variables, operationName, raw };
}


src/parse-body.ts
import type { IncomingMessage } from "http";
import type { Gunzip, Inflate } from "zlib";
import zlib from "zlib";
import querystring from "querystring";
import getBody from "raw-body";
import httpError from "http-errors";
import contentType from "content-type";

import type { ParsedMediaType } from "content-type";

export interface GraphQLParams {
  query: string | null;
  variables: { readonly [name: string]: unknown } | null;
  operationName: string | null;
  raw: boolean;
}

export type Request = IncomingMessage & { body?: unknown };

// Return a decompressed stream, given an encoding.
function decompressed(
  req: Request,
  encoding: string
): Request | Inflate | Gunzip {
  switch (encoding) {
    case "identity":
      return req;
    case "deflate":
      return req.pipe(zlib.createInflate());
    case "gzip":
      return req.pipe(zlib.createGunzip());
  }
  throw httpError(415, `Unsupported content-encoding "${encoding}".`);
}

/**
 * RegExp to match an Object-opening brace "{" as the first non-space
 * in a string. Allowed whitespace is defined in RFC 7159:
 *
 *     ' '   Space
 *     '\t'  Horizontal tab
 *     '\n'  Line feed or New line
 *     '\r'  Carriage return
 */
const jsonObjRegex = /^[ \t\n\r]*\{/u;

// Read and parse a request body.
async function readBody(
  req: Request,
  typeInfo: ParsedMediaType
): Promise<string> {
  const charset = typeInfo.parameters.charset?.toLowerCase() ?? "utf-8";

  // Assert charset encoding per JSON RFC 7159 sec 8.1
  if (!charset.startsWith("utf-")) {
    throw httpError(415, `Unsupported charset "${charset.toUpperCase()}".`);
  }

  // Get content-encoding (e.g. gzip)
  const contentEncoding = req.headers["content-encoding"];
  const encoding =
    typeof contentEncoding === "string"
      ? contentEncoding.toLowerCase()
      : "identity";
  const length = encoding === "identity" ? req.headers["content-length"] : null;
  const limit = 100 * 1024; // 100kb
  const stream = decompressed(req, encoding);

  // Read body from stream.
  try {
    return await getBody(stream, { encoding: charset, length, limit });
  } catch (rawError: unknown) {
    const error = httpError(
      400,
      /* istanbul ignore next: Thrown by underlying library. */
      rawError instanceof Error ? rawError : String(rawError)
    );

    error.message =
      error.type === "encoding.unsupported"
        ? `Unsupported charset "${charset.toUpperCase()}".`
        : `Invalid body: ${error.message}.`;
    throw error;
  }
}

/**
 * Provided a "Request" provided by express or connect (typically a node style
 * HTTPClientRequest), Promise the body data contained.
 */
export async function parseBody(
  req: IncomingMessage,
  body?: unknown
): Promise<Record<string, unknown>> {
  // If express has already parsed a body as a keyed object, use it.
  if (typeof body === "object" && !(body instanceof Buffer)) {
    return body === null ? {} : (body as Record<string, unknown>);
  }

  // Skip requests without content types.
  if (req.headers["content-type"] === undefined) {
    return {};
  }

  const typeInfo = contentType.parse(req);

  // If express has already parsed a body as a string, and the content-type
  // was application/graphql, parse the string body.
  if (typeof body === "string" && typeInfo.type === "application/graphql") {
    return { query: body };
  }

  // Already parsed body we didn't recognise? Parse nothing.
  if (body) {
    return {};
  }

  const rawBody = await readBody(req, typeInfo);
  // Use the correct body parser based on Content-Type header.
  switch (typeInfo.type) {
    case "application/graphql":
      return { query: rawBody };
    case "application/json":
      if (jsonObjRegex.test(rawBody)) {
        try {
          return JSON.parse(rawBody);
        } catch {
          // Do nothing
        }
      }
      throw httpError(400, "POST body sent invalid JSON.");
    case "application/x-www-form-urlencoded":
      return querystring.parse(rawBody);
  }

  // If no Content-Type header matches, parse nothing.
  return {};
}


src/graphql-param.ts
import type { IncomingMessage } from "http";
import { URLSearchParams } from "url";
import httpError from "http-errors";

import { parseBody } from "./parse-body";

export interface GraphQLParams {
  query: string | null;
  variables: { readonly [name: string]: unknown } | null;
  operationName: string | null;
  raw: boolean;
}

export async function getGraphQLParams(
  request: IncomingMessage,
  body?: unknown
): Promise<GraphQLParams> {
  const urlData = new URLSearchParams(request.url?.split("?")[1]);
  const bodyData = await parseBody(request, body);

  // GraphQL Query string.
  let query = urlData.get("query") ?? (bodyData.query as string | null);
  if (typeof query !== "string") {
    query = null;
  }

  // Parse the variables if needed.
  let variables = (urlData.get("variables") ?? bodyData.variables) as {
    readonly [name: string]: unknown;
  } | null;
  if (typeof variables === "string") {
    try {
      variables = JSON.parse(variables);
    } catch {
      throw httpError(400, "Variables are invalid JSON.");
    }
  } else if (typeof variables !== "object") {
    variables = null;
  }

  // Name of GraphQL operation to execute.
  let operationName =
    urlData.get("operationName") ?? (bodyData.operationName as string | null);
  if (typeof operationName !== "string") {
    operationName = null;
  }

  const raw = urlData.get("raw") !== null || bodyData.raw !== undefined;

  return { query, variables, operationName, raw };
}


src/parse-query.ts
import {
  execute,
  getOperationAST,
  parse,
  Source,
  specifiedRules,
  validate,
  validateSchema
} from "graphql";
import httpError from "http-errors";

import { GraphQLParams } from "./graphql-param";
import { OptionsData } from "./option";

import type { Context } from "koa";
import type { DocumentNode, ExecutionResult } from "graphql";

type Result = { result: ExecutionResult; document?: DocumentNode };

export const parseQuery = async (
  data: OptionsData,
  params: GraphQLParams,
  ctx: Context
): Promise<Result> => {
  const { query, variables, operationName } = params;
  const { request, response } = ctx;
  const { rootValue, schema, validationRules = [], fieldResolver } = data;

  // If there is no query, but GraphiQL will be displayed, do not produce
  // a result, otherwise return a 400: Bad Request.
  if (!query) {
    throw httpError(400, "Must provide query string.");
  }

  // Validate Schema
  const schemaValidationErrors = validateSchema(schema);
  if (schemaValidationErrors.length > 0) {
    // Return 500: Internal Server Error if invalid schema.
    response.status = 500;
    return { result: { errors: schemaValidationErrors } };
  }
  // GraphQL source.
  const source = new Source(query, "GraphQL request");

  let document: DocumentNode;
  // Parse source to AST, reporting any syntax error.
  try {
    document = parse(source);
  } catch (syntaxError) {
    // Return 400: Bad Request if any syntax errors errors exist.
    response.status = 400;
    return { result: { errors: [syntaxError] } };
  }

  const rules = specifiedRules.concat(validationRules);

  // Validate AST, reporting any errors.
  const validationErrors = validate(schema, document, rules);
  if (validationErrors.length > 0) {
    // Return 400: Bad Request if any validation errors exist.
    response.status = 400;
    return { result: { errors: validationErrors } };
  }

  // Only query operations are allowed on GET requests.
  if (request.method === "GET") {
    // Determine if this GET request will perform a non-query.
    const operationAST = getOperationAST(document, operationName);
    if (operationAST && operationAST.operation !== "query") {
      // Otherwise, report a 405: Method Not Allowed error.
      response.set("Allow", "POST");
      throw httpError(
        405,
        `Can only perform a ${operationAST.operation} operation from a POST request.`
      );
    }
  }

  // Perform the execution, reporting any errors creating the context.
  try {
    const result = await execute(
      schema,
      document,
      rootValue,
      ctx,
      variables,
      operationName,
      fieldResolver
    );
    response.status = 200;
    return { result, document };
  } catch (contextError) {
    // Return 400: Bad Request if any execution context errors exist.
    response.status = 400;
    return { result: { errors: [contextError] } };
  }
};
